### PR TITLE
Add SquiggleCop to baseline analyzer settings

### PR DESCRIPTION
**Pull request by MattKotsenas**
Add [SquiggleCop](https://github.com/MattKotsenas/SquiggleCop) to baseline the analyzers and their configuration. This intends to prevent accidental changes / regressions to the project, such as #165.

Changes made
- Add `SquiggleCop.Tasks` and wire into the existing `$(PedanticMode)` to validate baselines
- Add `SquiggleCop.Tool` to support ad-hoc baselining / troubleshooting
- Update `CONTRIBUTING.md` with re-baselining procedure
- Update SDK in `global.json` to latest version and disable rollforward to support reproducible build
- Update pipelines to upload SARIF files to aid debugging

### CODE REVIEW

**Comment by MattKotsenas** [comment]
⚠️ This PR is a rebase on-top of 165. Once that PR lands we should rebase those changes away. ⚠️

**Review by rjmurillo — APPROVED** [review]

File: build/targets/codeanalysis/CodeAnalysis.targets [hunk 1]
```diff
@@ -3,5 +3,22 @@
     <PedanticMode Condition=" '$(PedanticMode)' == '' ">$([MSBuild]::ValueOrDefault('$(ContinuousIntegrationBuild)', 'false'))</PedanticMode>
     <TreatWarningsAsErrors>$(PedanticMode)</TreatWarningsAsErrors>
     <MSBuildTreatWarningsAsErrors>$(PedanticMode)</MSBuildTreatWarningsAsErrors>
+    <SquiggleCop_Enabled>$(PedanticMode)</SquiggleCop_Enabled>
   </PropertyGroup>
+
+ <Target Name="SetErrorLog" BeforeTargets="CoreCompile">
```
[the remark this answers is not in the document]
Comment (MattKotsenas) [reply build/targets/codeanalysis/CodeAnalysis.targets]: Fixed

File: build/targets/codeanalysis/CodeAnalysis.targets [hunk 2]
```diff
@@ -3,5 +3,22 @@
...
+     ErrorLog is needed for SquiggleCop.
+
+     The value is set in a Target and not directly as a property because `$(IntermediateOutputPath)` and `$(OutputPath)`
+     are calculated properties and thus shouldn't be relied on during the initial property evaluation phase.
+     See https://github.com/dotnet/sdk/issues/41852.
+
+     We use `$(IntermediateOutputPath)` to ensure the file ends up in the `obj/` folder and not with sources to clearly
+     delineate inputs and outputs.
+   -->
+   <PropertyGroup Condition=" '$(ErrorLog)' == '' ">
```
[the remark this answers is not in the document]
Comment (MattKotsenas) [reply build/targets/codeanalysis/CodeAnalysis.targets]: Fixed

File: build/targets/codeanalysis/CodeAnalysis.targets [hunk 3]
```diff
@@ -3,5 +3,22 @@
...
+     <ErrorLog>$(IntermediateOutputPath)/$(MSBuildProjectName).sarif,version=2.1</ErrorLog>
+   </PropertyGroup>
+ </Target>
```
[the remark this answers is not in the document]
Comment (MattKotsenas) [reply build/targets/codeanalysis/CodeAnalysis.targets]: Fixed

File: global.json [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {
   "sdk": {
-    "version": "8.0.101",
-    "rollForward": "latestFeature"
+    "version": "8.0.303",
+    "rollForward": "disable"
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply global.json:4]: Had an offline discussion with @MattKotsenas 

There's a gotcha here. See https://github.com/MattKotsenas/SquiggleCop/blob/main/README.md. TL;DR - `rollForward` behavior can introduce new analyzers (8.0.400 includes a new one), so we could limit it to `latestPatch`. However, there's issue https://github.com/actions/setup-dotnet/issues/448 where the GitHub action doesn't support it. The only `rollForward` strategy that currently does what it says on the tin is `disabled`.

 This change will be extracted to its own and documented

[the remark this answers is not in the document]
Comment (rjmurillo) [reply global.json:4]: This is taken care of in #170. Thanks @MattKotsenas !

**Review by rjmurillo — APPROVED** [review]
You want to wait for #170 and rebase?

**Comment by MattKotsenas** [comment]
> You want to wait for #170 and rebase?

Yes, once that one merges I'll rebase away the changes here and merge.

**Pull request merged**